ipfs/ipfs.go

```diff
@@ -0,0 +1,316 @@
+			}
+		}(peerInfo)
+	}
+	wg.Wait()
+	return nil
+}
+
+func getUnixfsFile(path string) (files.File, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	st, err := file.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := files.NewReaderPathFile(path, file, st)
+	if err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+func getUnixfsNode(path string) (files.Node, error) {
+	st, err := os.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := files.NewSerialFile(path, false, st)
+	if err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+type UploadResponse struct {
+	Status int
+	Cid    string
+	Path   string
+}
+
+// External Functions
+func UploadData(ctx context.Context, data []byte, node iface.CoreAPI) (UploadResponse, error) {
+	// use --only-hash option for preprocessor
+	var permissions uint32 = 0644 // or whatever you need
+	err := os.WriteFile(tempDir, data, fs.FileMode(permissions))
+	if err != nil {
+		return UploadResponse{Status: status.StatusInternalServerError}, err
+	}
+
+	// TODO revisit the file strategy and see if there is a better way
+	file, err := getUnixfsNode(tempDir)
+	if err != nil {
+		return UploadResponse{Status: status.StatusInternalServerError}, err
+	}
+
+	cidFile, err := node.Unixfs().Add(ctx, file)
+	if err != nil {
+		return UploadResponse{Status: status.StatusInternalServerError}, err
+	}
+
+	fmt.Printf("Added file to IPFS with CID %s\n", cidFile.String())
+
+	newLocation := ipfsStorageDir
+	os.Mkdir(newLocation, 0755)
+	if err != nil {
+		return UploadResponse{Status: status.StatusInternalServerError}, err
+	}
+	newLocation += cidFile.String()
+	os.Mkdir(newLocation, 0755)
+	if err != nil {
+		return UploadResponse{Status: status.StatusInternalServerError}, err
+	}
+	newLocation += "/bytes.txt"
+	os.WriteFile(newLocation, data, 0644)
+
+	return UploadResponse{
+		Status: status.StatusOK,
+		Cid:    cidFile.String(),
+	}, nil
+}
+
+type DownloadResponse struct {
+	Status    int
+	FileData  []byte
+	NameSpace string
+	Path      string
+}
+
+func DownloadData(ctx context.Context, cid string, node iface.CoreAPI) (DownloadResponse, error) {
+	outputBasePath, err := ioutil.TempDir("", "example")
+	if err != nil {
+		panic(fmt.Errorf("could not create output dir (%v)", err))
+	}
+
+	//TODO strip cid down to just the cid string portion
+
+	path := outputBasePath + cid
+	cidPath := ifacepath.New(cid)
+
+	//check local file system first
+	if _, err := os.Stat(ipfsStorageDir + "/" + cid + "/bytes.txt"); !errors.Is(err, os.ErrNotExist) {
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			return DownloadResponse{Status: status.StatusInternalServerError}, err
+		}
+		return DownloadResponse{
+			Status:    status.StatusOK,
+			FileData:  data,
+			Path:      cidPath.String(),
+			NameSpace: cidPath.Namespace(),
+		}, nil
+	}
+
+	rootNodeFile, err := node.Unixfs().Get(ctx, cidPath)
```

Comment:
are we checking the cidPath for validity before operating on the FS?

Reply:
There is some validity checks first thing in UnicFs().Get() if they fail then the reason why will be returned in err.

Reply:
Resolving.